src/Simulation/QsharpFoundation/Diagnostics/Assert.qs

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Quantum.Diagnostics {  
+    /// # Summary
+    /// Asserts that measuring the given qubits in the given Pauli basis will
+    /// always have the given result.
+    ///
+    /// # Input
+    /// ## bases
+    /// A measurement effect to assert the probability of, expressed as a
+    /// multi-qubit Pauli operator.
+    /// ## qubits
+    /// A register on which to make the assertion.
+    /// ## result
+    /// The expected result of `Measure(bases, qubits)`.
+    /// ## msg
+    /// A message to be reported if the assertion fails.
+    ///
+    /// # Remarks
+    /// Note that the Adjoint and Controlled versions of this operation will not
+    /// check the condition.
+    ///
+    /// # See Also
+    /// - Microsoft.Quantum.Diagnostics.AssertMeasurementProbability
+    operation AssertMeasurement(bases : Pauli[], qubits : Qubit[], result : Result, msg : String) : Unit
```

Comment:
To support the new layout I'm trying to introduce where QsharpCore is based on QsharpFoundation, I moved these intrinsic methods here. @cgranade, does this seem reasonable?